File: application/models/traction_index.php
<?php

class Traction_index extends CI_Model{
	
	private $fields = array('facebook_share_count', 'seomoz_authority', 'twitter_url_count', 'twitter_url_mentions', 'twitter_followers');
	private $field_aggrs;

	function __construct()
	{
		// Call the Model constructor
		parent::__construct();
		
		$this->field_aggrs = $this->get_field_aggrs();
	}
	
	private function get_field_aggrs(){
		$this->db->select('type, MIN(data) as min, MAX(data) as max', false);
		$this->db->where_in('type', $this->fields);
		$this->db->group_by('type');
		$aggr_results = $this->db->get('app_external_data');
		$type_aggr = array();
		foreach($aggr_results->result_array() as $result){
			$type_aggr[$result['type']] = $result;
		}
		
		return $type_aggr;
	}
	
	function get_traction_index($app_id){
		if(empty($this->field_aggrs)){
			$this->field_aggrs = $this->get_field_aggrs();
		}
		
		$sub_indices = array();
		
		$this->db->select('type, data, time_added');
		$this->db->where('time_added >', '(CURDATE() - INTERVAL 1 MONTH)', false);
		$this->db->where('app_id', $app_id);
		$this->db->where_in('type', $this->fields);
		$this->db->group_by('type');
		$results = $this->db->get('app_external_data');
		
		foreach($results->result_array() as $app_field){
			if(empty($app_field['data']) || !is_numeric($app_field['data'])) continue;
			
			$sub_indices[$app_field['type']] = (log($app_field['data']) - log($this->field_aggrs[$app_field['type']]['min'])) / (log($this->field_aggrs[$app_field['type']]['max']) - log($this->field_aggrs[$app_field['type']]['min']));
		}
		
		if(count($sub_indices) == 0){
			return false;
		}
		
		$traction_value = pow(array_product($sub_indices), (1 / count($sub_indices)));
		
		$traction_index = round($traction_value * 100);
		
		return $traction_index;
	}
	
}

File: python/lib/traction_index.py
from lib import app_model
import math
from pprint import pprint

class TractionIndex:

	fields = (
		'facebook_share_count',
		'seomoz_authority',
		'compete_unique_visitors',
		'alexa_pageviews_per_million'
	)

	field_aggrs = {}

	def __init__(self):
		self.get_field_aggrs()

	def get_field_aggrs(self):
		field_aggrs = app_model.get_field_aggrs(self.fields)
		for field_aggr in field_aggrs:
			self.field_aggrs[field_aggr.type] = {'min':field_aggr.min, 'max':field_aggr.max}

		return True

	def get_traction_index(self, app_id):
		if not self.field_aggrs:
			self.get_field_aggrs()

		sub_indices = {}

		app_external_data = app_model.get_app_external_data(self.fields, app_id)
		for app_data in app_external_data:
			if not app_data.data:
				continue

			sub_index_val = (math.log(app_data.data) - math.log(self.field_aggrs[app_data.type]['min'])) / (math.log(self.field_aggrs[app_data.type]['max']) - math.log(self.field_aggrs[app_data.type]['min']))
			if sub_index_val > 0:
				sub_indices[app_data.type] = sub_index_val

		if len(sub_indices) == 0:
			return False

		sub_indices_product = 1
		for sub_index_type, sub_index_val in sub_indices.items():
			sub_indices_product *= sub_index_val
		traction_value = math.pow(sub_indices_product, (1 / float(len(sub_indices))))
		traction_index = round(traction_value * 100)

		return traction_index